File: Talleres RESFTUL JAVA DOTNET/Java/api-java-bank/src/main/java/com/allpasoft/apijavabank/application/service/TransactionService.java
package com.allpasoft.apijavabank.application.service;

import com.allpasoft.apijavabank.application.dto.*;
import com.allpasoft.apijavabank.domain.entity.Account;
import com.allpasoft.apijavabank.domain.entity.Transaction;
import com.allpasoft.apijavabank.domain.entity.TransactionType;
import com.allpasoft.apijavabank.domain.repository.AccountRepository;
import com.allpasoft.apijavabank.domain.repository.TransactionRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class TransactionService {

    private final TransactionRepository transactionRepository;
    private final AccountRepository accountRepository;

    public TransactionService(TransactionRepository transactionRepository, AccountRepository accountRepository) {
        this.transactionRepository = transactionRepository;
        this.accountRepository = accountRepository;
    }

    public Optional<TransactionDto> getById(Long id) {
        return transactionRepository.findById(id).map(this::toDto);
    }

    public List<TransactionDto> getByAccountId(Long accountId) {
        return transactionRepository.findByAccountIdOrDestinationAccountId(accountId).stream()
                .map(this::toDto)
                .collect(Collectors.toList());
    }

    public List<TransactionDto> getAll() {
        return transactionRepository.findAllByOrderByCreatedAtDesc().stream()
                .map(this::toDto)
                .collect(Collectors.toList());
    }

    @Transactional
    public Optional<TransactionDto> deposit(CreateDepositDto dto) {
        Optional<Account> accountOpt = accountRepository.findById(dto.getAccountId());
        if (accountOpt.isEmpty() || !accountOpt.get().getIsActive()) {
            return Optional.empty();
        }

        Account account = accountOpt.get();
        account.setBalance(account.getBalance().add(dto.getAmount()));
        account.setUpdatedAt(LocalDateTime.now());
        accountRepository.save(account);

        Transaction transaction = new Transaction();
        transaction.setTransactionType(TransactionType.DEPOSIT);
        transaction.setAmount(dto.getAmount());
        transaction.setDescription(dto.getDescription() != null ? dto.getDescription() : "Depósito");
        transaction.setAccountId(dto.getAccountId());
        transaction.setCreatedAt(LocalDateTime.now());

        Transaction saved = transactionRepository.save(transaction);
        saved.setAccount(account);
        return Optional.of(toDto(saved));
    }

    @Transactional
    public Optional<TransactionDto> withdraw(CreateWithdrawalDto dto) {
        Optional<Account> accountOpt = accountRepository.findById(dto.getAccountId());
        if (accountOpt.isEmpty() || !accountOpt.get().getIsActive()) {
            return Optional.empty();
        }

        Account account = accountOpt.get();
        if (account.getBalance().compareTo(dto.getAmount()) < 0) {
            return Optional.empty();
        }

        account.setBalance(account.getBalance().subtract(dto.getAmount()));
        account.setUpdatedAt(LocalDateTime.now());
        accountRepository.save(account);

        Transaction transaction = new Transaction();
        transaction.setTransactionType(TransactionType.WITHDRAWAL);
        transaction.setAmount(dto.getAmount());
        transaction.setDescription(dto.getDescription() != null ? dto.getDescription() : "Retiro");
        transaction.setAccountId(dto.getAccountId());
        transaction.setCreatedAt(LocalDateTime.now());

        Transaction saved = transactionRepository.save(transaction);
        saved.setAccount(account);
        return Optional.of(toDto(saved));
    }

    @Transactional
    public Optional<TransactionDto> transfer(CreateTransferDto dto) {
        Optional<Account> sourceOpt = accountRepository.findById(dto.getSourceAccountId());
        Optional<Account> destOpt = accountRepository.findById(dto.getDestinationAccountId());

        if (sourceOpt.isEmpty() || !sourceOpt.get().getIsActive()) {
            return Optional.empty();
        }
        if (destOpt.isEmpty() || !destOpt.get().getIsActive()) {
            return Optional.empty();
        }

        Account source = sourceOpt.get();
        Account dest = destOpt.get();

        if (source.getBalance().compareTo(dto.getAmount()) < 0) {
            return Optional.empty();
        }

        source.setBalance(source.getBalance().subtract(dto.getAmount()));
        source.setUpdatedAt(LocalDateTime.now());
        dest.setBalance(dest.getBalance().add(dto.getAmount()));
        dest.setUpdatedAt(LocalDateTime.now());

        accountRepository.save(source);
        accountRepository.save(dest);

        Transaction transaction = new Transaction();
        transaction.setTransactionType(TransactionType.TRANSFER);
        transaction.setAmount(dto.getAmount());
        transaction.setDescription(dto.getDescription() != null ? dto.getDescription() : "Transferencia");
        transaction.setAccountId(dto.getSourceAccountId());
        transaction.setDestinationAccountId(dto.getDestinationAccountId());
        transaction.setCreatedAt(LocalDateTime.now());

        Transaction saved = transactionRepository.save(transaction);
        saved.setAccount(source);
        saved.setDestinationAccount(dest);
        return Optional.of(toDto(saved));
    }

    private TransactionDto toDto(Transaction transaction) {
        return new TransactionDto(
                transaction.getId(),
                transaction.getTransactionType().ordinal(),
                transaction.getAmount(),
                transaction.getDescription(),
                transaction.getAccountId(),
                transaction.getAccount() != null ? transaction.getAccount().getAccountNumber() : "",
                transaction.getDestinationAccountId(),
                transaction.getDestinationAccount() != null ? transaction.getDestinationAccount().getAccountNumber() : null,
                transaction.getCreatedAt()
        );
    }
}

File: Talleres RESFTUL JAVA DOTNET/Dotnet/API-DOTNET-BANK/Application/Service/TransactionService.cs
using API_DOTNET_BANK.Application.DTOs.Transaction;
using API_DOTNET_BANK.Application.Interface;
using API_DOTNET_BANK.Domain.Entities;
using API_DOTNET_BANK.Domain.Enums;
using API_DOTNET_BANK.Domain.Repository;

namespace API_DOTNET_BANK.Application.Service
{
    public class TransactionService : ITransactionService
    {
        private readonly ITransactionRepository _transactionRepository;
        private readonly IAccountRepository _accountRepository;

        public TransactionService(ITransactionRepository transactionRepository, IAccountRepository accountRepository)
        {
            _transactionRepository = transactionRepository;
            _accountRepository = accountRepository;
        }

        public async Task<TransactionDto?> GetByIdAsync(int id)
        {
            var transaction = await _transactionRepository.GetByIdAsync(id);
            return transaction != null ? ToDto(transaction) : null;
        }

        public async Task<IEnumerable<TransactionDto>> GetByAccountIdAsync(int accountId)
        {
            var transactions = await _transactionRepository.GetByAccountIdAsync(accountId);
            return transactions.Select(ToDto);
        }

        public async Task<IEnumerable<TransactionDto>> GetAllAsync()
        {
            var transactions = await _transactionRepository.GetAllAsync();
            return transactions.Select(ToDto);
        }

        public async Task<TransactionDto?> DepositAsync(CreateDepositDto dto)
        {
            var account = await _accountRepository.GetByIdAsync(dto.AccountId);
            if (account == null || !account.IsActive) return null;

            account.Balance += dto.Amount;
            await _accountRepository.UpdateAsync(account);

            var transaction = new Transaction
            {
                TransactionType = TransactionType.Deposit,
                Amount = dto.Amount,
                Description = dto.Description ?? "Depósito",
                AccountId = dto.AccountId,
                CreatedAt = DateTime.UtcNow
            };

            var created = await _transactionRepository.CreateAsync(transaction);
            created.Account = account;
            return ToDto(created);
        }

        public async Task<TransactionDto?> WithdrawAsync(CreateWithdrawalDto dto)
        {
            var account = await _accountRepository.GetByIdAsync(dto.AccountId);
            if (account == null || !account.IsActive) return null;
            if (account.Balance < dto.Amount) return null;

            account.Balance -= dto.Amount;
            await _accountRepository.UpdateAsync(account);

            var transaction = new Transaction
            {
                TransactionType = TransactionType.Withdrawal,
                Amount = dto.Amount,
                Description = dto.Description ?? "Retiro",
                AccountId = dto.AccountId,
                CreatedAt = DateTime.UtcNow
            };

            var created = await _transactionRepository.CreateAsync(transaction);
            created.Account = account;
            return ToDto(created);
        }

        public async Task<TransactionDto?> TransferAsync(CreateTransferDto dto)
        {
            var sourceAccount = await _accountRepository.GetByIdAsync(dto.SourceAccountId);
            var destAccount = await _accountRepository.GetByIdAsync(dto.DestinationAccountId);
            
            if (sourceAccount == null || !sourceAccount.IsActive) return null;
            if (destAccount == null || !destAccount.IsActive) return null;
            if (sourceAccount.Balance < dto.Amount) return null;

            sourceAccount.Balance -= dto.Amount;
            destAccount.Balance += dto.Amount;
            
            await _accountRepository.UpdateAsync(sourceAccount);
            await _accountRepository.UpdateAsync(destAccount);

            var transaction = new Transaction
            {
                TransactionType = TransactionType.Transfer,
                Amount = dto.Amount,
                Description = dto.Description ?? "Transferencia",
                AccountId = dto.SourceAccountId,
                DestinationAccountId = dto.DestinationAccountId,
                CreatedAt = DateTime.UtcNow
            };

            var created = await _transactionRepository.CreateAsync(transaction);
            created.Account = sourceAccount;
            created.DestinationAccount = destAccount;
            return ToDto(created);
        }

        private static TransactionDto ToDto(Transaction transaction) => new()
        {
            Id = transaction.Id,
            TransactionType = (int)transaction.TransactionType,
            Amount = transaction.Amount,
            Description = transaction.Description,
            AccountId = transaction.AccountId,
            AccountNumber = transaction.Account?.AccountNumber ?? "",
            DestinationAccountId = transaction.DestinationAccountId,
            DestinationAccountNumber = transaction.DestinationAccount?.AccountNumber,
            CreatedAt = transaction.CreatedAt
        };
    }
}
